pulp_docker/tests/functional/api/test_recursive_add.py

```diff
@@ -0,0 +1,140 @@
+            {'repository': self.to_repo['_href'], 'content_units': [ml_i]})
+        latest_version_href = self.client.get(self.to_repo['_href'])['_latest_version_href']
+        latest = self.client.get(latest_version_href)
+
+        # No tags added
+        self.assertFalse('docker.manifest-tag' in latest['content_summary']['added'])
+        # 1 manifest list 2 manifests
+        self.assertEqual(latest['content_summary']['added']['docker.manifest']['count'], 3)
+        self.assertEqual(latest['content_summary']['added']['docker.manifest-blob']['count'], 4)
+
+    def test_tagged_manifest_list_recursion(self):
+        """Add a tagged manifest list, and its related manifests and blobs."""
+        ml_i = self.client.get("{unit_path}?{filters}".format(
+            unit_path=DOCKER_TAG_PATH,
+            filters="name=ml_i",
+        ))['results'][0]['_href']
+        self.client.post(
+            DOCKER_RECURSIVE_ADD_PATH,
+            {'repository': self.to_repo['_href'], 'content_units': [ml_i]})
+        latest_version_href = self.client.get(self.to_repo['_href'])['_latest_version_href']
+        latest = self.client.get(latest_version_href)
+        # 1 manifest list 2 manifests
+        self.assertEqual(latest['content_summary']['added']['docker.manifest-tag']['count'], 1)
+        self.assertEqual(latest['content_summary']['added']['docker.manifest']['count'], 3)
+        self.assertEqual(latest['content_summary']['added']['docker.manifest-blob']['count'], 4)
+
+    def test_tagged_manifest_recursion(self):
+        """Add a tagged manifest and its related blobs."""
+        manifest_a = self.client.get("{unit_path}?{filters}".format(
+            unit_path=DOCKER_TAG_PATH,
+            filters="name=manifest_a",
+        ))['results'][0]['_href']
+        self.client.post(
+            DOCKER_RECURSIVE_ADD_PATH,
+            {'repository': self.to_repo['_href'], 'content_units': [manifest_a]})
+        latest_version_href = self.client.get(self.to_repo['_href'])['_latest_version_href']
+        latest = self.client.get(latest_version_href)
+
+        self.assertEqual(latest['content_summary']['added']['docker.manifest-tag']['count'], 1)
+        self.assertEqual(latest['content_summary']['added']['docker.manifest']['count'], 1)
+        self.assertEqual(latest['content_summary']['added']['docker.manifest-blob']['count'], 2)
+
+    def test_tag_replacement(self):
+        """Add a tagged manifest to a repo with a tag of that name already in place."""
+        manifest_a = self.client.get("{unit_path}?{filters}".format(
+            unit_path=DOCKER_TAG_PATH,
+            filters="name=manifest_a",
+        ))['results'][0]['_href']
+        self.client.post(
+            DOCKER_RECURSIVE_ADD_PATH,
+            {'repository': self.to_repo['_href'], 'content_units': [manifest_a]})
+        self.client.post(
+            DOCKER_RECURSIVE_ADD_PATH,
+            {'repository': self.to_repo['_href'], 'content_units': [manifest_a]})
```

Comment:
NOTE: This is the same tag. The test would be better to use a different tag, but that requires https://github.com/pulp/pulp_docker/pull/394 to be merged.